### PR TITLE
feat(platform/github): flag to control whether PRs can be edited by maintainers if `forkToken`is set

### LINKED ISSUE #16657: Fork mode: Previously edited branch does not get overridden/deleted
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.119.0

### Please select which platform you are using if self-hosting.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

Github

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

With fork mode, renovate stops updating a group when a PR has been manually modified previously. e.g.

1. Renovate sends an update for a group called `angular-deps`
2. The PR starts to fail because we have size failures.
3. A maintainer pushes the pull request to fix up the size failures (this is possible since the PR has allowed maintainers to modify the fork branch)
4. The PR gets merged.
5. A few days/weeks later, still no update is sent because Renovate still has the old branch that was modified by someone other than the Renovate Git user.

This issue was visible in the `angular/components` repository, and here is the self-hosted runner entry: https://github.com/angular/dev-infra/runs/7415118482?check_suite_focus=true (look for `Last commit author does not match git author email`)

The PR modified in (3) was not immortal and the branch never got deleted upon merge: https://github.com/angular/components/pull/25187 (this PR is an example -- Note that the branch was deleted manually just today after discovering this)

**Note**: It's unclear to me how this is supposed to work. Should the branch be deleted on merge? This is obviously more difficult when Renovate operates in a fork. Or should we just disable maintainer edits when `forkMode: true`?

Also super happy to look into contributing a fix if the expected behavior is clear and it's not a super large undertaking.

### Relevant debug logs

<details><summary>Logs</summary>
DEBUG: getBranchPr(ng-renovate/angular-shared-dev-infra-code) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, undefined, open) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, undefined, closed) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: branchExists=true (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: dependencyDashboardCheck=undefined (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: recreateClosed is false (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, build: update angular shared dev-infra code to cc061c7, !open) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: findPr(renovate/angular-shared-dev-infra-code, build: update angular shared dev-infra code to cc061c7, !open) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: prAlreadyExisted=false (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: Checking if PR has been edited (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: Last commit author does not match git author email - branch has been modified (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
       "branchName": "ng-renovate/angular-shared-dev-infra-code",
       "lastAuthor": "paulgschwendtner@gmail.com",
       "gitAuthorEmail": "angular-robot@google.com"
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, undefined, !open) (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: Found PR #25180 (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
DEBUG: Found old PR but the SHA is different (repository=angular/components, branch=ng-renovate/angular-shared-dev-infra-code)
       "oldPrNumber": 25180,
       "oldPrSha": "f5d8e2acdc03b25318d6c78e8c9a420071d0c3eb",
       "branchSha": "f21a665004c41f68cff5effb8ad2c7b1217e728f"
DEBUG: getBranchPr(ng-renovate/angular-shared-dev-infra-code) (repository=angular/components)
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, undefined, open) (repository=angular/components)
DEBUG: findPr(ng-renovate/angular-shared-dev-infra-code, undefined, closed) (repository=angular/components)
DEBUG: Ensuring Dependency Dashboard (repository=angular/components)
DEBUG: ensureIssue(Dependency Dashboard) (repository=angular/components)
DEBUG: Issue is open and up to date - nothing to do (repository=angular/components)
</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by devversion**
## Changes

A new option called `forkModeAllowMaintainerEdits` is introduced. This option is only relevant if Renovate is used with `forkToken`.
* The option is by default `true`, keeping backwards compatibilty
* The option will only have an effect for `github`. 

The option decides whether maintainer edits are allowed to edit pull requests.

## Context

If a maintainer pushes changes to a fork renovate PR, the branch will never be deleted by Renovate because there are changes not corresponding to the Git account configured in Renovate.

This prevents future updates as Renovate basically locks down the branch and marks it as `pr-edited`, preventing it from being overridden or cleaned-up as part of the scheduled branch deletion.

We can fix this by encouraging that maintainers cherry-pick the commit and make changes in a separate PR. Users of the experimental fork mode option can do this by setting `prAllowMaintainerEdits = false`. It is not possible for maintainers to trivially delete the branch of the e.g. fork robot account because they would need to log into that and not every team member would e.g. necessarily have access to the robot account.

- Fixes #16657

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [X] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/github/index.ts
Comment (rarkins) [inline lib/modules/platform/github/index.ts:1497]: It had been my assumption until now that this feature was quite useful for users. Therefore I want to think whether this change is necessary or if there's "another way".

Comment (devversion) [reply lib/modules/platform/github/index.ts:1497]: yeah, it totally is. I think in the issue I was advocating for a simpler branch deletion policy, like:

* If a PR is for the onboarding branch, never delete it
* If a PR is modified
   * only delete if there is no associated PR

but it seemed like changing the process here would be a rather larger undertaking, while for forked PRs this is already an improvement compared to renovate suddenly not updating a certain dependency at all anymore

Comment (viceice) [reply lib/modules/platform/github/index.ts:1497]: this is a breaking change, so it should better be configurable and enabled by default

Comment (devversion) [reply lib/modules/platform/github/index.ts:1497]: yeah, I can make it configurable. Though could also see it as a fix because the current "intended behavior" does not work and has more significant consequences (like people not knowing that they miss updates after having edited a fork branch)

Comment (viceice) [reply lib/modules/platform/github/index.ts:1497]: doesn't the delete branch after merge function (GitHub repo settings) solve the issue too?

Comment (devversion) [reply lib/modules/platform/github/index.ts:1497]: Interesting idea! (didn't know this existed). It looks like it wouldn't be able to delete branches from the fork account though. e.g. consider https://github.com/angular/dev-infra/pull/821 where the PR branch is not an upstream branch
Screenshot 2023-01-10 at 21 23 35

We had the setting enabled but it didn't delete the branch. Likely a maintainer cannot delete a branch from a user fork

Comment (rarkins) [reply lib/modules/platform/github/index.ts:1497]: I agree with https://github.com/renovatebot/renovate/pull/19771#discussion_r1066218416

**Review by viceice — CHANGES_REQUESTED** [review]
needs to be configurable and enabled by default to be not breaking

**Review by rarkins — CHANGES_REQUESTED** [review]
1. This is only relevant to forkingRenovate. Right now it's configurable for both and the docs don't make it clear enough. Let's try to avoid confusing the 99% of users not affected by this
2. Can this be made global config instead (i.e. controlled by the bot admin, not the repo admin)?

File: lib/config/options/index.ts
Comment (HonkingGoose) [inline lib/config/options/index.ts:1561]:
```suggestion
      'Decides if maintainers can edit Renovate pull requests.',
```

How about something like this?

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2184]:
```suggestion
Use `prAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests.
```

How about something like this?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md
Comment (devversion) [reply docs/usage/configuration-options.md:2184]: thanks!

**Comment by devversion** [comment]
I've addressed all feedback (and appreciate anybody having another look. Thanks for all the comments)

<sup>Github seems to just send it to one of the reviewers</sup>

File: lib/config/options/index.ts
Comment (viceice) [inline lib/config/options/index.ts:1564]: so only the bot owner can change it? how would this PR help with forking renovate app?

File: lib/modules/platform/github/index.ts
Comment (viceice) [inline lib/modules/platform/github/index.ts:1497]: this should be moved back to the if condition below, as it only works with forkToken.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts
Comment (devversion) [reply lib/config/options/index.ts:1564]: good point. should we go back to make it a repository setting?

Comment (devversion) [reply lib/config/options/index.ts]: I've switched it back to a per-repo setting since the forking renovate GitHub app is a good argument. Since you also want this to be configurable, this seems like the only option.

**Comment by devversion** [comment]
I've incorporated all the docs suggestions and switched back to a repo config setting. A global option would simplify this a bit, but would make it impossible to use this option when e.g. the forking renovate GitHub app is used.

Sorry. I've also force-pushed to clean-up the commits which made it more difficult to review- a full review is easier in this situation anyway (and very likely needed anyway)

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2189]:
```suggestion
When Renovate runs in a fork, this allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md
Comment (devversion) [reply docs/usage/configuration-options.md]: updated (sorry was OOO until today)

**Review by viceice — DISMISSED** [review]

**Review by HonkingGoose — DISMISSED** [review]

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:2192]:
```suggestion
## forkModeAllowMaintainerEdits

Use `forkModeAllowMaintainerEdits` to control if maintainers can edit Renovate's pull requests when in fork mode.

When Renovate runs in a fork, this allows project maintainers to make manual changes to the Renovate PR branch, without needing to create another new PR.

<!-- prettier-ignore -->
!!! note
    This option is only relevant if you set `forkToken`.
```


This will require reordering within the file too

File: lib/config/options/index.ts
Comment (rarkins) [inline lib/config/options/index.ts:1559]:
```suggestion
    name: 'forkModeAllowMaintainerEdits',
```

File: lib/config/types.ts
Comment (rarkins) [inline lib/config/types.ts:233]:
```suggestion
  forkModeAllowMaintainerEdits?: boolean;
```

File: lib/workers/repository/config-migration/pr/index.ts
Comment (rarkins) [inline lib/workers/repository/config-migration/pr/index.ts:103]:
```suggestion
        allowMaintainerEdits: config.forkModeAllowMaintainerEdits,
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts
Comment (devversion) [reply lib/config/options/index.ts:1563]: @viceice brought up a good point on this. Users of e.g. the Renovate forking app would not be able to benefit from this option then.

**Comment by devversion** [comment]
I'm good restricting the option name also to `forkMode`. Updated as per review suggestions. One extra note: I changed the `createPr` option from `allowMaintainerEdits` also to `forkModeAllowMaintainerEdits` because otherwise it would be quite ambiguous. We could even go further and move it into platform options I assume.

**Comment by devversion** [comment]
FYI: I will wait with any docs changes until we have settled on runtime code, naming, or if this should be a global only flag etc.

File: docs/usage/configuration-options.md
Comment (devversion) [reply docs/usage/configuration-options.md:886]: The `a` is intentionally omitted because it should refer to the `forkToken` option. I think adding `a` would make this a little less clear.

Comment (HonkingGoose) [reply docs/usage/configuration-options.md:886]: It felt a bit wrong to me. 😄 I'll defer to the judgment of a native English speaker on this. 😉

Comment (devversion) [reply docs/usage/configuration-options.md:886]: me too 😄

File: lib/workers/repository/onboarding/pr/index.ts
Comment (viceice) [inline lib/workers/repository/onboarding/pr/index.ts:177]: @rarkins i think we should pass this via the platform options below?

Comment (devversion) [reply lib/workers/repository/onboarding/pr/index.ts:177]: yeah, I was wondering too (https://github.com/renovatebot/renovate/pull/19771#issuecomment-1426039302). 

will make the change based on what you decide

Comment (devversion) [reply lib/workers/repository/onboarding/pr/index.ts:177]: friendly ping so I can make all the suggested changes

Comment (rarkins) [reply lib/workers/repository/onboarding/pr/index.ts:177]: Yes, please pass it as a platform option

File: docs/usage/configuration-options.md
Comment (devversion) [reply docs/usage/configuration-options.md:882]: I've tried to add more info here, but it's a little odd because the old workflow is basically *broken* 😄

File: docs/usage/configuration-options.md
Comment (viceice) [inline docs/usage/configuration-options.md:903]:
```suggestion
Allowing maintainers to edit pull requests directly is helpful when Renovate pull requests require additional changes.
The reviewer can simply push to the pull request without having to create a new PR. [More details here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
```
one sentence per line

File: docs/usage/configuration-options.md
Comment (viceice) [inline docs/usage/configuration-options.md:905]:
```suggestion
You may decide to disallow edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657).
If this option is enabled, reviewers will need to create a new PR if additional changes are needed.
```

File: lib/config/options/index.ts
Comment (viceice) [inline lib/config/options/index.ts:1938]:
```suggestion
    default: false,
```
so we don't break existing users

File: lib/modules/platform/types.ts
Comment (viceice) [inline lib/modules/platform/types.ts:95]:
```suggestion
  forkModeDisallowMaintainerEdits?: boolean;
```
use same name as config option

File: lib/config/options/index.ts
Comment (devversion) [reply lib/config/options/index.ts:1938]: good catch. I reversed the option with the recent changs

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:897]:
```suggestion
## forkModeBlockMaintainerEdits
```

**Review by viceice — APPROVED** [review]

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:899]:
```suggestion
Use `forkModeBlockMaintainerEdits` to block maintainers from editing Renovate's pull requests when in fork mode.
```

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:906]:
```suggestion
You may decide to block edits to Renovate pull requests in order to workaround issues in Renovate where modified fork branches are not deleted properly: [See this issue](https://github.com/renovatebot/renovate/issues/16657).
```

File: lib/config/options/index.ts
Comment (HonkingGoose) [inline lib/config/options/index.ts:1933]:
```suggestion
    name: 'forkModeBlockMaintainerEdits',
```

File: lib/config/options/index.ts
Comment (HonkingGoose) [inline lib/config/options/index.ts:1935]:
```suggestion
      'Block maintainer pushes to Renovate pull requests when running in fork mode.',
```

File: lib/config/types.ts
Comment (HonkingGoose) [inline lib/config/types.ts:233]:
```suggestion
  forkModeBlockMaintainerEdits?: boolean;
```

File: lib/modules/platform/github/index.spec.ts
Comment (HonkingGoose) [inline lib/modules/platform/github/index.spec.ts:2221]:
```suggestion
            forkModeBlockMaintainerEdits: false,
```

File: lib/modules/platform/github/index.spec.ts
Comment (HonkingGoose) [inline lib/modules/platform/github/index.spec.ts:2266]:
```suggestion
            forkModeBlockMaintainerEdits: true,
```

File: lib/modules/platform/github/index.ts
Comment (HonkingGoose) [inline lib/modules/platform/github/index.ts:1503]:
```suggestion
      platformOptions?.forkModeBlockMaintainerEdits !== true;
```

File: lib/modules/platform/types.ts
Comment (HonkingGoose) [inline lib/modules/platform/types.ts:95]:
```suggestion
  forkModeBlockMaintainerEdits?: boolean;
```

File: lib/workers/repository/update/pr/index.ts
Comment (HonkingGoose) [inline lib/workers/repository/update/pr/index.ts:53]:
```suggestion
    forkModeBlockMaintainerEdits: config.forkModeBlockMaintainerEdits,
```

**Review by HonkingGoose — COMMENTED** [review]
How about using a simpler name like `forkModeBlockMaintainerEdits`?

The word `disallow` feels clunky to me. I think a word like `block` is easier to understand.

I'll let the maintainers decide if they want to apply my suggestions.

**Comment by devversion** [comment]
Yeah, if the maintainers think they prefer `block`, I can make the change. I personally prefer `(dis)allow` since that is the terminology used by GitHub as well. Blocking feels different to me (but I'm a non-native speaker) See:

image

and https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork.

**Comment by viceice** [comment]
i think disallow is the right thing 🤷‍♂️

**Comment by HonkingGoose** [comment]
> I personally prefer `(dis)allow` since that is the terminology used by GitHub as well.

Good point, let's stick with `dis(allow)` then. 😄

**Review by HonkingGoose — APPROVED** [review]

**Comment by HonkingGoose** [comment]
@viceice can you resolve my PR review comments? I can't do this myself: [^docs]

> You can resolve a conversation in a pull request if you opened the pull request or if you have write access to the repository where the pull request was opened.

[^docs]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#resolving-conversations

**Review by rarkins — APPROVED** [review]

**Pull request merged**
